### ISSUE TITLE
[Feature Request] Send heartbeat details to next activity invocation

### ISSUE THREAD

**Issue by lorensr** [temporalio/samples-typescript#101]
As described here: https://typescript.temporal.io/api/classes/activity.context/#heartbeat

Add to this sample: https://github.com/temporalio/samples-typescript/blob/main/activities-cancellation-heartbeating/src/workflows.ts

Thanks @mabadir for reporting

**Comment by lorensr** [comment]
From the docs, I imaginge it would be something like:

```ts
export async function runCancellableActivity(): Promise<void> {
  try {
    await fakeProgress();
  } catch (err) {
    if (err instanceof ActivityFailure && err.cause instanceof TimeoutFailure) {
      await fakeProgress(err.cause.lastHeartbeatDetails)
    } else {
      throw err;
    }
  }
}
```

https://typescript.temporal.io/api/classes/client.timeoutfailure/#lastheartbeatdetails

**Comment by sw-yx** [comment]
i think the design pattern we need to work out is how to loop this activity until completion, handling multiple failures until an expected timeout. 

as a step 2, perhaps this would then become a first class Activity API where the SDK would handle it for the user instead of having to write all this heartbeat boilerplate

**Comment by Sushisource** [comment]
I think the much easier way to solve this problem is to set an appropriate retry policy. If you let the activity fail, and then say "oh but I actually wanted to pass the progress on and invoke it again", well, it seems like your retry policy wasn't really what you meant. What you actually wanted here was for it to just keep retrying.

**Comment by mjameswh** [comment]
Fixed in #102
